critique.py

```diff
@@ -115,13 +126,106 @@ def find_NX_class_nodes(self, parent, nx_class = 'NXentry'):
+            MAGIC_HDF4 = b'\x0e\x03\x13\x01\x00\xc8\x00\x00'
+            with open(os.path.join(self.path, self.fname), "rb") as file:
+                #file.seek(0)
+                sig = file.read(8)
+            if sig == MAGIC_HDF4:
+                self.filetype = "HDF4"
+        return self.filetype
+
+#    def test_01a(self, path, fname):
+#        if path is None and fname is None:
+#            return "Header"
+#        with open(os.path.join(self.path, self.fname), "rb") as file:
+#            file.seek(0)
+#            sig = file.read(8)
+#        return ":".join("{:02x}".format(x) for x in bytearray(sig))
+
+    def test_02(self, path, fname):
+        if path is None and fname is None:
+            return "NXentry Count"
+        if self.filetype == "HDF5":
+            with h5py.File(os.path.join(self.path, self.fname), mode="r") as root:
+                NXentry_nodes = self.find_NX_class_nodes(root, "NXentry")
+                self.nNXentry = len(NXentry_nodes)
+            if self.nNXentry == 0:
+                return "not NeXus"
+        elif self.filetype == "XML":
+            self.nNXentry = "*"
+        elif self.filetype == "HDF4":
+            self.nNXentry = "*"
+        else:
+            self.nNXentry = "-"
+        return self.nNXentry
+
+    def test_03(self, path, fname):
+        if path is None and fname is None:
+            return "Application Def's"
+        if self.filetype == "HDF5":
+            if self.nNXentry < 1:
+                AppDefList = "-"
+            else:
+                ad_list = set() # like a list, but only keep unique strings
+                with h5py.File(os.path.join(self.path, self.fname), mode="r") as root:
+                    NXentry_nodes = self.find_NX_class_nodes(root, "NXentry")
+                    for entry in NXentry_nodes:
+                        subentry_list = self.find_NX_class_nodes(entry, "NXsubentry")
+                        if len(subentry_list) == 0:
+                            if 'definition' in list(entry):
+                                ad_list.add(str(entry['definition'][0], encoding='utf-8'))
+                        else:
+                            for sub in subentry_list:
+                                ad_list.add(str(sub['definition'][0], encoding='utf-8'))
+                if len(ad_list) == 0:
+                    AppDefList = "None found"
+                else:
+                    AppDefList = ",".join(ad_list)
+        elif self.filetype == "XML":
+            AppDefList = "*"
+            
+        elif self.filetype == "HDF4":
+            AppDefList = "*"
+            
+        else:
+            AppDefList = "-"
+            
+        return AppDefList
 
 
 class Registrar(object):
     '''keep track of critiqued files in an internal dictionary'''
 
     def __init__(self):
         self.db = {}
+        self.test_bank = [func for func in dir(Critic) if callable(getattr(Critic, func)) and func.startswith("test_")]
```

Comment:
While efficient for execution, list comprehensions such as this can be tricky to review and edit.  Also, the line length becomes a bit long.  I often break a list comprehension such as this into several lines to improve readability without compromising efficiency:

```
self.test_bank = [
    func
    for func in dir(Critic)
    if callable(getattr(Critic, func)) and func.startswith("test_")
]
```